File: app/src/main/java/com/friendoye/rss_reader/model/RssParser.java
package com.friendoye.rss_reader.model;

import android.graphics.Bitmap;
import android.util.Log;
import android.util.Xml;

import com.nostra13.universalimageloader.core.ImageLoader;

import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;
import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;
import java.io.InputStream;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;

/**
 * Base class for all parsers of RSS feed and detailed link.
 */
abstract public class RssParser {
    public static final String PARSE_EXCEPTION_TAG = "ParseException";

    public List<RssFeedItem> parseRssStream(InputStream input)
            throws XmlPullParserException, IOException {
        try {
            XmlPullParser parser = Xml.newPullParser();
            parser.setInput(input, null);
            parser.next();
            return reedItems(parser);
        } finally {
            if (input != null) {
                input.close();
            }
        }
    }

    abstract protected RssFeedItem reedItem(XmlPullParser parser)
            throws XmlPullParserException, IOException;

    abstract public String retrieveDescription(Document doc)
            throws RuntimeException;

    public Bitmap retrieveLargeImage(Document doc)
            throws RuntimeException {Elements blocks = doc.select("meta[property=\"og:image\"]");
        try {
            String imageLink = blocks.get(0).attr("content");
            if (imageLink != null) {
                return ImageLoader.getInstance().loadImageSync(imageLink);
            } else {
                throw new RuntimeException("No link in tag!");
            }
        } catch (NullPointerException e) {
            throw new RuntimeException("No tag was found. Info: " + e);
        }
    }

    /**
     * ============== HELP METHODS ====================
     */

    /**
     * Return reversed list of RssFeedItem from XML.
     */
    protected List<RssFeedItem> reedItems(XmlPullParser parser)
            throws XmlPullParserException, IOException {
        LinkedList<RssFeedItem> items = new LinkedList<>();

        parser.require(XmlPullParser.START_TAG, null, "rss");
        while (skipUntil(parser, "item")) {
            parser.require(XmlPullParser.START_TAG, null, "item");
            RssFeedItem item = reedItem(parser);
            //item.source = mSource;
            items.addFirst(item);
        }

        return items;
    }

    protected String readText(XmlPullParser parser, String tagName)
            throws XmlPullParserException, IOException {
        parser.require(XmlPullParser.START_TAG, null, tagName);
        String title = retrieveText(parser);
        parser.require(XmlPullParser.END_TAG, null, tagName);
        return title;
    }

    protected Date readDate(XmlPullParser parser, String tagName)
            throws XmlPullParserException, IOException {
        parser.require(XmlPullParser.START_TAG, null, tagName);
        String stringDate = retrieveText(parser);
        Date date = null;;
        try {
            final SimpleDateFormat formatter =
                    new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss",
                            Locale.ENGLISH);
            date = formatter.parse(stringDate);
        } catch (ParseException e) {
            Log.i(PARSE_EXCEPTION_TAG,
                    "readDate(): invalid parser. Info: " + e);
        }
        parser.require(XmlPullParser.END_TAG, null, tagName);
        return date;
    }

    protected String readImageUrl(XmlPullParser parser, String tagName)
            throws XmlPullParserException, IOException {
        parser.require(XmlPullParser.START_TAG, null, tagName);
        String imageUrl = parser.getAttributeValue(null, "url");
        parser.next();
        parser.require(XmlPullParser.END_TAG, null, tagName);
        return imageUrl;
    }

    protected String retrieveText(XmlPullParser parser)
            throws IOException, XmlPullParserException {
        String result = "";
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.getText();
            parser.nextTag();
        }
        return result;
    }

    protected void skipCurrentTag(XmlPullParser parser)
            throws XmlPullParserException, IOException {
        int depth = 1;
        while (depth != 0) {
            switch (parser.next()) {
                case XmlPullParser.END_TAG:
                    depth--;
                    break;
                case XmlPullParser.START_TAG:
                    depth++;
                    break;
            }
        }
    }

    protected boolean skipUntil(XmlPullParser parser, String targetName)
            throws XmlPullParserException, IOException {
        while (parser.next() != XmlPullParser.END_DOCUMENT) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }
            String tagName = parser.getName();
            if (tagName.equals(targetName)) {
                return true;
            }
        }
        return false;
    }

}

File: app/src/main/java/com/friendoye/rss_reader/model/RssParser.kt
package com.friendoye.rss_reader.model

import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.util.Log
import coil.ImageLoader.Companion.invoke
import android.util.Xml
import coil.ImageLoader
import coil.request.GetRequest
import com.friendoye.rss_reader.Application
import org.jsoup.nodes.Document
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import java.io.IOException
import java.io.InputStream
import java.lang.IndexOutOfBoundsException
import java.lang.NullPointerException
import java.lang.RuntimeException
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

/**
 * Base class for all parsers of RSS feed and detailed link.
 */
abstract class RssParser {

    @Throws(XmlPullParserException::class, IOException::class)
    fun parseRssStream(input: InputStream?): List<RssFeedItem> {
        return try {
            val parser = Xml.newPullParser()
            parser.setInput(input, null)
            parser.next()
            reedItems(parser)
        } finally {
            input?.close()
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    protected abstract fun reedItem(parser: XmlPullParser): RssFeedItem?

    @Throws(RuntimeException::class)
    abstract fun retrieveDescription(doc: Document): String?

    @Throws(RuntimeException::class)
    suspend fun retrieveLargeImage(doc: Document): Bitmap? {
        val blocks = doc.select("meta[property^=\"og:image\"]")
        try {
            val imageLink = blocks[0].attr("content")
            if (imageLink != null) {
                // TODO: Replace with Coil
                val request: GetRequest = GetRequest.Builder(Application.getInstance())
                    .data(imageLink)
                    .allowHardware(false) // Disable hardware bitmaps.
                    .build()
                return (ImageLoader(Application.getInstance())
                    .execute(request)
                    .drawable as? BitmapDrawable)
                    ?.bitmap
            } else {
                Log.e(PARSE_EXCEPTION_TAG, "No link in tag!")
            }
        } catch (e: NullPointerException) {
            Log.e(PARSE_EXCEPTION_TAG, "No tag was found. Info: $e")
        } catch (e: IndexOutOfBoundsException) {
            Log.e(PARSE_EXCEPTION_TAG, "No tag was found. Info: $e")
        }
        return null
    }

    /**
     * ============== HELP METHODS ====================
     */

    /**
     * Return reversed list of RssFeedItem from XML.
     */
    @Throws(XmlPullParserException::class, IOException::class)
    protected fun reedItems(parser: XmlPullParser): List<RssFeedItem> {
        val items = LinkedList<RssFeedItem>()
        parser.require(XmlPullParser.START_TAG, null, "rss")
        while (skipUntil(parser, "item")) {
            parser.require(XmlPullParser.START_TAG, null, "item")
            val item = reedItem(parser)
            if (item != null) {
                items.addFirst(item) // This method should be implemented
            }
        }
        return items
    }

    @Throws(XmlPullParserException::class, IOException::class)
    protected fun readText(parser: XmlPullParser, tagName: String?): String {
        parser.require(XmlPullParser.START_TAG, null, tagName)
        val title = retrieveText(parser)
        parser.require(XmlPullParser.END_TAG, null, tagName)
        return title
    }

    @Throws(XmlPullParserException::class, IOException::class)
    protected fun readDate(parser: XmlPullParser, tagName: String?): Date? {
        parser.require(XmlPullParser.START_TAG, null, tagName)
        val stringDate = retrieveText(parser)
        var date: Date? = null
        try {
            val formatter = SimpleDateFormat(
                "EEE, dd MMM yyyy HH:mm:ss",
                Locale.ENGLISH
            )
            date = formatter.parse(stringDate)
        } catch (e: ParseException) {
            Log.i(
                PARSE_EXCEPTION_TAG,
                "readDate(): invalid parser. Info: $e"
            )
        }
        parser.require(XmlPullParser.END_TAG, null, tagName)
        return date
    }

    @Throws(XmlPullParserException::class, IOException::class)
    protected fun readImageUrl(parser: XmlPullParser, tagName: String?): String {
        parser.require(XmlPullParser.START_TAG, null, tagName)
        val imageUrl = parser.getAttributeValue(null, "url")
        parser.next()
        parser.require(XmlPullParser.END_TAG, null, tagName)
        return imageUrl
    }

    @Throws(IOException::class, XmlPullParserException::class)
    protected fun retrieveText(parser: XmlPullParser): String {
        var result = ""
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.text
            parser.nextTag()
        }
        return result
    }

    @Throws(XmlPullParserException::class, IOException::class)
    protected fun skipCurrentTag(parser: XmlPullParser) {
        var depth = 1
        while (depth != 0) {
            when (parser.next()) {
                XmlPullParser.END_TAG -> depth--
                XmlPullParser.START_TAG -> depth++
            }
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    protected fun skipUntil(parser: XmlPullParser, targetName: String): Boolean {
        while (parser.next() != XmlPullParser.END_DOCUMENT) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            val tagName = parser.name
            if (tagName == targetName) {
                return true
            }
        }
        return false
    }

    companion object {
        const val PARSE_EXCEPTION_TAG = "ParseException"
    }
}